### ISSUE TITLE
Blake 3 Support

### ISSUE THREAD

**Issue by travankor** [openssl/openssl#11613]
Can you add Blake3 support to openssl? It's over 10x faster than sha2 and will probably be widely adopted.

**Comment by romen** [comment]
As a general rule, a necessary condition to include a new algorithm into the main code base is that the algorithm should be part of a standard from a recognized standardization body.

This is not to say that we won't encourage/assist the community to work on a pull request to add support for it, but that most likely such feature wouldn't be merged until a standard that uses it is available.

That being said, with the new Provider concept, it should be more easy than ever to seamlessly add support through an external module for an algorithm that is not yet listed among the ones provided by the built-in providers.
Writing an external provider that adds support for Blake3 could be an ideal way of working towards a contribution to the project once the necessary conditions are met.

**Comment by romen** [comment]
I'm also going to add the `good first issue` label.

Working on a dedicated provider is not what I would call a "bite-sized" endeavor, and it's not for beginners, but I still believe it could be an ideal starting step for an experienced newcomer that wants to understand how to plug new functionality into `libcrypto`.

**Comment by 4cad** [comment]
@paulidale I would be interested in taking a crack at this. I am not experienced with openssl though - when you are talking about the new architecture, are you referring to something like https://github.com/openssl/openssl/blob/master/providers/implementations/digests/blake2b_prov.c ?

**Comment by paulidale** [comment]
The _new architecture_ was meant to be a reference to how much of the surrounding OpenSSL code has changed.  You are on the right path though.  That file implements the digest, there is some additional wrapping in `blake2_prov.c` in the same directory.  For blake3, it should be a separate source file & I'd just include everything there (i.e. don't create a `blake3_prov.c`).

To implement blake3, you will need to do it in its own provider (because it doesn't yet meet the minimum standard for inclusion into OpenSSL).  This will require a blake3 provider (look at `providers/default.c` and `providers/null.c`) for details.  The latter is a minimal provider that offers nothing, the former indicates how to offer algorithms, you only want digests so it will be one table with one entry.

The building infrastructure should be straightforward enough: compile the blake3_provider, the blake3 implementation, link into a loadable module.  It is fine to link against libcrypto.so.  The result should be a provider supplying blake3.

Once blake3 is accepted as a standard, incorporation in OpenSSL would be straightforward:

- put the blake3.c source into `providers/implementations/digests` (hopefully as is);
- add the blake3 entry to `providers/default.c` (again as is) and
- modify the `providers/implementations/digests/build.info` to include it.

**Comment by 4cad** [comment]
Perfect, thanks that seems strait-forward enough although I am not 100% clear on the pieces that is why I wanted to give it a go. Why don't I get some first draft code working first and the details should become clear in the comments of a PR

**Comment by mattcaswell** [comment]
As noted by @paulidale blake3 support won't be currently acceptable into OpenSSL because there  is no standard. You will have to maintain it as an external provider based implementation in some other repo until such time as it does become standardised.

**Comment by 4cad** [comment]
Right, so to summarize an unofficial external Blake3 provider will be created in some other repo that depends only on libcrypto with the goals of

1. Before Blake3 is standardized, people who want bleeding edge will be able to use the unofficial and unsupported external Blake3 provider.
2. After Blake3 is standardized, assuming the algorithm has not changed much it should easy to simply copy the external provider into the OpenSSL default provider where it will then go through the usual review process.

Seems reasonable to me.

**Comment by mattcaswell** [comment]
Exactly.

**Comment by J-Montgomery** [comment]
I also found this issue interesting and took it on as a weekend project when it looked like @4cad hadn't touched it in a couple months. [Repo is here](https://github.com/J-Montgomery/blake3-prov). It passes all the fixed output-length test vectors from the reference implementation.

I wasn't able to test extended length output because of a minor issue with the provider API. BLAKE3 has the strange feature that it supports an unlimited length output, which can be set when handling OSSL_FUNC_DIGEST_FINAL. I think it'd be best to implement this as a settable context param, but OpenSSL doesn't have the functionality to do this or expose it to library users. It only checks OSSL_FUNC_DIGEST_GET_PARAMS, where the provider has to return the default output length. I've left the code in there to handle it regardless and am open to other suggestions on how to do it.

I'm not sure where we should go from here though. Should it be documented anywhere besides this ticket?

**Comment by paulidale** [comment]
What about `OSSL_DIGEST_PARAM_XOFLEN`?  This is used in the SHAKE implementation to set the output length.

**Comment by J-Montgomery** [comment]
Don't know how I missed that, but it completely solves the issue. Thanks

**Comment by mattcaswell** [comment]
Nice work @J-Montgomery!

**Comment by odiferousmint** [comment]
Any progress on this? Pointers? I would like to see BLAKE3 in OpenSSL, or rather, Erlang. I do not want Elixir bindings to the Rust code, in all fairness.

**Comment by t8m** [comment]
> Any progress on this? Pointers? I would like to see BLAKE3 in OpenSSL, or rather, Erlang. I do not want Elixir bindings to the Rust code, in all fairness.

It is very unlikely for Blake3  to be included into OpenSSL in 3.0. However you have a link above to the Blake 3 provider that will be usable with OpenSSL 3.0.

**Comment by odiferousmint** [comment]
Thank you!

**Comment by Silur** [comment]
Blake3 reached version 1.0 on it's reference implementation today - https://github.com/BLAKE3-team/BLAKE3
I don't know whether that qualifies as a "standard implementation" or not, but with version 1 in my opinion it's about time to consider inclusion,

**Comment by mattcaswell** [comment]
> I don't know whether that qualifies as a "standard implementation"

No, it does not qualify. As noted by @romen above:

> a necessary condition to include a new algorithm into the main code base is that the algorithm should be part of a standard from a recognized standardization body.

**Comment by dralley** [comment]
What standard is Blake2b part of, given that it is already merged?  I spent a few minutes looking, and couldn't find one.

I found https://www.rfc-editor.org/rfc/rfc7693.html, but that explicitly states "This document is not an Internet Standards Track specification; it is published for informational purposes."

**Comment by t8m** [comment]
Informational RFCs are sufficient for adopting an algorithm by OpenSSL.

**Comment by bricss** [comment]
I was wondering if it's possible for OpenSSL to adopt `Blake3` after all, with humongous billboard warning ⚠️ that it is not approved by any institution? 🤔

**Comment by t8m** [comment]
At least an informational RFC would be necessary.

**Comment by nhorman** [comment]
I still don't see an RFC for Blake3 available, so this seems like its not going to get planned.

Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by oconnor663** [comment]
@veorq did our informational RFC submission go through? Last I remember there were some technical issues with the submission system.

**Comment by bkmgit** [comment]
@oconnor663 @veorq  It seems not to have been posted https://datatracker.ietf.org/submit/status/142299/

**Comment by nhorman** [comment]
@dralley sure, can I assume from the above that you, @bkmgit or @oconnor663  or @travankor would be willing to submit a PR for this?  I'm happy to assign it to one of you as a community issue

**Comment by bkmgit** [comment]
Probably @dralley @oconnor663 or @travankor could implement this faster, but am happy to do it.

**Comment by Bronek** [comment]
> @oconnor663 @veorq It seems not to have been posted https://datatracker.ietf.org/submit/status/142299/

it was resubmitted on July 22 : https://datatracker.ietf.org/doc/draft-aumasson-blake3/

**Comment by bkmgit** [comment]
@nhorman Thanks

**Comment by t8m** [comment]
A question for OpenSSL management (BACs) is whether a specification in https://github.com/C2SP/C2SP would be sufficient to include the algorithm.

**Comment by veorq** [comment]
Thanks @t8m for your response, and sorry about the duplicate ticket. 

For the context: we started an [IETF I-D](https://datatracker.ietf.org/doc/draft-aumasson-blake3/), but the discussions were inconclusive and we were told that C2SP could be enough for OpenSSL (cc @filosottile)

**Comment by FiloSottile** [comment]
> A question for OpenSSL management (BACs) is whether a specification in https://github.com/C2SP/C2SP would be sufficient to allow including the algorithm.

As a data point to inform this decision, here is [the formal Chrome CT Log Policy](https://googlechrome.github.io/CertificateTransparency/log_policy.html) referencing a C2SP specification.

**Comment by bkmgit** [comment]
IETF CFRG is unwilling to go through the review process as they indicate SHA3 should be sufficient for most use cases - https://mailarchive.ietf.org/arch/browse/cfrg/?=blake

However, it is possible to send a draft to the independent stream for the purposes of having a documented reference.

**Comment by bkmgit** [comment]
Example of an informational  draft https://datatracker.ietf.org/doc/html/rfc6234

**Comment by oconnor663** [comment]
I don't think this got explicitly mentioned above, so for context: a [specification of BLAKE3](https://github.com/C2SP/C2SP/blob/main/BLAKE3.md) was added to C2SP [back in September](https://github.com/C2SP/C2SP/pull/104).
